### PR TITLE
Expand WeakEventManager to Support Delegate

### PR DESCRIPTION
**Pull request by TheCodeTraveler**
## Description of Change

Implements #2703 

This PR expands `WeakEventManager` to support `Delegate`.

Currently, `Microsoft.Maui.Controls.WeakEventManger` only supports `event EventHandler`. It does not currently work with the other [`event`](https://docs.microsoft.com/en-us/dotnet/standard/events) types in C#, like [`Delegate`](https://docs.microsoft.com/en-us/dotnet/standard/events/?WT.mc_id=mobile-43581-bramin#delegates).

This means that `Microsoft.Maui.Controls.WeakEventManger` does not currently work for the most common `Delegate` event, [`INotifyPropertyChanged.PropertyChanged`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged.propertychanged?WT.mc_id=mobile-43581-bramin).

```cs
readonly WeakEventManager weakEventManager = new WeakEventManager();

event PropertyChangedEventHandler INotifyPropertyChanged.PropertyChanged
{
    add => weakEventManager.AddEventHandler(value); // This PR fixes this compiler error
    remove => weakEventManager.RemoveEventHandler(value);  // This PR fixes this compiler error
}

void OnPropertyChanged([CallerMemberName] in string propertyName = "") => weakEventManager.HandleEvent(this, new PropertyChangedEventArgs(propertyName), nameof(INotifyPropertyChanged.PropertyChanged));
```

The Xamarin Community Toolkit has included this [expanded WeakEventManager](https://docs.microsoft.com/xamarin/community-toolkit/helpers/delegateweakeventmanager?WT.mc_id=mobile-43581-bramin) (named `DelegateWeakEventManager` to avoid naming conflicts) since v1.0.0-pre4 released on October 21, 2020.

### WeakEventManager.cs
- Enables nullable (`#nullable enable`)
  - Required to support `INotifyPropertyChanged`
    - [`INotifyPropertyChanged.PropertyChanged` event is nullable](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged.propertychanged?view=net-6.0&WT.mc_id=mobile-43581-bramin) as of .NET 5.0
      ```cs
      event System.ComponentModel.PropertyChangedEventHandler? PropertyChanged;
      ```
    - Screenshot of the null reference error if nullable is not enabled on WeakEventManager:
        image

- Updates `AddEventHandler`
  - **Previous**
    ```cs
    public void AddEventHandler(EventHandler handler, [CallerMemberName] string eventName = null)
    ```
  - **Updated**
    ```cs
    public void AddEventHandler(Delegate? handler, [CallerMemberName] string eventName = "")
    ```
    - Note: `Delegate?` must be nullable to support `INotifyPropertyChanged` 
- Updates `RemoveEventHandler`
  - **Previous**
    ```cs
    public void RemoveEventHandler(EventHandler handler, [CallerMemberName] string eventName = null)
    ```
  - **Updated**
    ```cs
    public void RemoveEventHandler(Delegate? handler, [CallerMemberName] string eventName = "")
    ```
    - Note: `Delegate?` must be nullable to support `INotifyPropertyChanged` 

### WeakEventManagerTests.cs 
- Enables nullable (`#nullable enable`)
- Adds `VerifyPropertyChanged` Test
  ```cs
  [Test]
  public void VerifyPropertyChanged()
  ```

### PR Checklist ###

- [x] Targets the correct branch 
- [x] Tests are passing (or failures are unrelated)
- [x] Targets a single property for a single control (or intertwined few properties)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might affect accessibility?
- [ ] Does this PR introduce a new control? (If yes, add an example using SemanticProperties to the SemanticsPage)
- [ ] APIs that modify focusability?
- [ ] APIs that modify any text property on a control?
- [ ] Does this PR modify view nesting or view arrangement in anyway?
- [ ] Is there the smallest possibility that your PR will change accessibility? 
- [ ] I'm not sure, please help me

If any of the above checkboxes apply to your PR, then the PR will need to provide testing to demonstrate that accessibility still works.

### CODE REVIEW

**Comment by brminnick** [comment]
Fixes #3253

**Review by Szymon2601 — APPROVED** [review]
image_

**Comment by Szymon2601** [comment]
> image_

**Review by hartez — APPROVED** [review]

**Comment by brminnick** [comment]
Looks like the test didn't fail, but rather timed-out for some reason

File: src/Controls/src/Core/WeakEventManager.cs [hunk 1]
```diff
@@ -87,7 +87,7 @@ public void RemoveEventHandler<TEventArgs>(EventHandler<TEventArgs> handler, [Ca
 			RemoveEventHandler(eventName, handler.Target, handler.GetMethodInfo());
 		}
 
-		public void RemoveEventHandler(EventHandler handler, [CallerMemberName] string eventName = null)
+		public void RemoveEventHandler(Delegate? handler, [CallerMemberName] string eventName = "")
```
Comment (dersia) [inline src/Controls/src/Core/WeakEventManager.cs:90]: I know this is already merged, but may I ask, what's the point of making it nullable, if you thrown if it is null?

Comment (brminnick) [reply src/Controls/src/Core/WeakEventManager.cs:90]: @dersa It’s to support `INotifyPropertyChanged`

Here’s the information I included on it from the PR:

> ### WeakEventManager.cs
> - Enables nullable (`#nullable enable`)
>   - Required to support `INotifyPropertyChanged`
>     - [`INotifyPropertyChanged.PropertyChanged` event is nullable](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged.propertychanged?view=net-6.0&WT.mc_id=mobile-43581-bramin) as of .NET 5.0
>       ```cs
>       event System.ComponentModel.PropertyChangedEventHandler? PropertyChanged;
>       ```
>     - Screenshot of the null reference error if nullable is not enabled on WeakEventManager:
>         image
>

**Pull request merged**